user/idtools_unix_test.go

```diff
@@ -0,0 +1,411 @@
+	tmpDir := t.TempDir()
+	setWorkingDirectory(t, tmpDir)
+
+	const expectedUIDGID = 101
+
+	for _, tc := range tests {
+		t.Run(tc.in, func(t *testing.T) {
+			for _, p := range tc.out {
+				_, err := os.Stat(p)
+				if !errors.Is(err, os.ErrNotExist) {
+					t.Fatalf("expected file not exists for %v, got %v", p, err)
+				}
+			}
+
+			if err := MkdirAllAndChown(tc.in, 0o755, expectedUIDGID, expectedUIDGID, WithOnlyNew); err != nil {
+				t.Fatal(err)
+			}
+
+			for _, p := range tc.out {
+				s := &unix.Stat_t{}
+				if err := unix.Stat(p, s); err != nil {
+					t.Error(err)
+					continue
+				}
+				if s.Uid != expectedUIDGID {
+					t.Errorf("expected UID: %d, got: %d", expectedUIDGID, s.Uid)
+				}
+				if s.Gid != expectedUIDGID {
+					t.Errorf("expected GID: %d, got: %d", expectedUIDGID, s.Gid)
+				}
+			}
+		})
+	}
+}
+
+// Change the current working directory for the duration of the test. This may
+// break if tests are run in parallel.
+func setWorkingDirectory(t *testing.T, dir string) {
+	t.Helper()
+	cwd, err := os.Getwd()
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Cleanup(func() {
+		if err := os.Chdir(cwd); err != nil {
+			t.Error(err)
+		}
+	})
+	if err = os.Chdir(dir); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestMkdirAndChown(t *testing.T) {
+	RequiresRoot(t)
+	dirName, err := os.MkdirTemp("", "mkdir")
+	if err != nil {
+		t.Fatalf("Couldn't create temp dir: %v", err)
+	}
+	defer os.RemoveAll(dirName)
+
+	testTree := map[string]node{
+		"usr": {0, 0},
+	}
+	if err := buildTree(dirName, testTree); err != nil {
+		t.Fatal(err)
+	}
+
+	// test a directory that already exists; should just chown to the requested uid/gid
+	if err := MkdirAndChown(filepath.Join(dirName, "usr"), 0o755, 99, 99); err != nil {
+		t.Fatal(err)
+	}
+	testTree["usr"] = node{99, 99}
+	verifyTree, err := readTree(dirName, "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	compareTrees(t, testTree, verifyTree)
+
+	// create a subdir under a dir which doesn't exist--should fail
+	if err := MkdirAndChown(filepath.Join(dirName, "usr", "bin", "subdir"), 0o755, 102, 102); err == nil {
+		t.Fatalf("Trying to create a directory with Mkdir where the parent doesn't exist should have failed")
+	}
+
+	// create a subdir under an existing dir; should only change the ownership of the new subdir
+	if err := MkdirAndChown(filepath.Join(dirName, "usr", "bin"), 0o755, 102, 102); err != nil {
+		t.Fatal(err)
+	}
+	testTree["usr/bin"] = node{102, 102}
+	verifyTree, err = readTree(dirName, "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	compareTrees(t, testTree, verifyTree)
+}
+
+func buildTree(base string, tree map[string]node) error {
+	for path, node := range tree {
+		fullPath := filepath.Join(base, path)
+		if err := os.MkdirAll(fullPath, 0o755); err != nil {
+			return fmt.Errorf("couldn't create path: %s; error: %w", fullPath, err)
+		}
+		if err := os.Chown(fullPath, node.uid, node.gid); err != nil {
+			return fmt.Errorf("couldn't chown path: %s; error: %w", fullPath, err)
```

Comment:
nit: no need to wrap `os` errors as they already have context (path and operation).